components/sources/cu_hesai/src/lib.rs

```diff
@@ -1 +1,263 @@
+}
+
+struct Xt32 {
+    socket: Socket,
+    addr: SocketAddr,
+    reftime: RefTime,
+    channel_elevations: [Angle; 32],
+}
+
+impl Xt32 {
+    /// This give the matching timestamps between the UTC time to the Robot time.
+    fn sync(&mut self, robot_clock: &RobotClock) {
+        self.reftime = (Utc::now(), robot_clock.now());
+    }
+}
+
+impl CuTaskLifecycle for Xt32 {
+    fn new(config: Option<&ComponentConfig>) -> CuResult<Self>
+    where
+        Self: Sized,
+    {
+        let addr: SocketAddr = if let Some(cfg) = config {
+            let addr_str = cfg.get("socket_addr").unwrap_or(DEFAULT_ADDR.to_string());
+            addr_str.as_str().parse().unwrap()
+        } else {
+            DEFAULT_ADDR.parse().unwrap()
+        };
+
+        let socket = Socket::new(Domain::IPV4, Type::DGRAM, Some(Protocol::UDP)).unwrap();
+        socket.bind(&SockAddr::from(addr)).unwrap();
+        socket.set_nonblocking(true).unwrap();
+
+        // just a temporary value, it will be redone at start.
+        let rt: RefTime = (Utc::now(), RobotClock::new().now());
+        Ok(Xt32 {
+            socket,
+            addr,
+            reftime: rt,
+            channel_elevations: generate_default_elevation_calibration(), // TODO: make the config able to override that
+        })
+    }
+    fn start(&mut self, robot_clock: &RobotClock) -> CuResult<()> {
+        self.sync(&robot_clock);
+        Ok(())
+    }
+}
+
+impl Freezable for Xt32 {}
+
+const MAX_POINTS: usize = 32 * 10;
+
+type LidarCuMsgPayload = LidarPayloadSoa<MAX_POINTS>;
+
+// In each round of firing, the firing sequence is from Channel 1 to Channel 32.
+// Assuming that the start time of Block 6 is t6, the laser firing time of Channel i is
+// t6 + [1.512µs * (i-1) + 0.28µs ], i∈{1, 2, ..., 32}.
+// in copper resolution as 1ns
+// t6 + 1512ns * (i-1) + 280ns
+fn channel_time(t6: CuTime, i: u64) -> CuTime {
+    if i == 0 {
+        CuDuration(t6.0 - 1512 + 280) // this is an underflow, so we just subtract the value
+    } else {
+        CuDuration(t6.0 + 1512 * (i - 1) + 280)
+    }
+}
+
+impl<'cl> CuSrcTask<'cl> for Xt32 {
+    type Output = output_msg!('cl, LidarCuMsgPayload);
+
+    fn process(&mut self, clock: &RobotClock, new_msg: Self::Output) -> CuResult<()> {
+        let payload = new_msg.payload_mut().insert(LidarCuMsgPayload::default());
+        let mut buf = [0u8; 1500];
+        match self.socket.read(&mut buf) {
+            Ok(size) => {
+                let lidar_packet = parser::parse_packet(&buf[..size])
+                    .map_err(|e| CuError::new_with_cause("Failed to parse Hesai UDP packet", e))?;
+
+                // this is the reference point for the block timings
+                let t6 = lidar_packet
+                    .block_ts(&self.reftime)
+                    .map_err(|e| CuError::new_with_cause("Failed to get block timings", e))?[5]; // 0 == channel 1, 5 == channel 6
+
+                let is_dual = lidar_packet.header.is_dual_return();
```

Comment:
Are we doing something this this information ?
How would we handle multiple returns in LidarCuMsgPayload ?

Reply:
ok let's drop that for that PR and we will add another one with a proper return order in the format and the drivers. thanks